### ISSUE TITLE
if there is no path, this algorithm cost so much time and give a wrong path

### ISSUE THREAD

**Issue by gouchaoer** [da-an/a-star#9]

**Comment by Xenoamor** [comment]
It's always going to be slow under a failure condition as it has to exhaust every single tile it has access to. The wrong path can be fixed by changing the tail end of `FindPath` to:
```
    CoordinateList path;
    // We have not got a valid path if our current coordinates do not equal the source
    if (current->coordinates == source_)
    {
        while (current != nullptr) {
            path.push_back(current->coordinates);
            current = current->parent;
        }
    }
```

This will return an empty path when it fails

**Comment by Ahmija** [comment]
> It's always going to be slow under a failure condition as it has to exhaust every single tile it has access to. The wrong path can be fixed by changing the tail end of `FindPath` to:
> 
> ```
>     CoordinateList path;
>     // We have not got a valid path if our current coordinates do not equal the source
>     if (current->coordinates == source_)
>     {
>         while (current != nullptr) {
>             path.push_back(current->coordinates);
>             current = current->parent;
>         }
>     }
> ```
> 
> This will return an empty path when it fails

it still gives a wrong path.

**Comment by BubsTheDev** [comment]
> It's always going to be slow under a failure condition as it has to exhaust every single tile it has access to. The wrong path can be fixed by changing the tail end of `FindPath` to:
> 
> ```
>     CoordinateList path;
>     // We have not got a valid path if our current coordinates do not equal the source
>     if (current->coordinates == source_)
>     {
>         while (current != nullptr) {
>             path.push_back(current->coordinates);
>             current = current->parent;
>         }
>     }
> ```
> 
> This will return an empty path when it fails

The original code sometimes gives a wrong path (the end point seems to get set to the edge of the graph/map maybe?). 
Unfortunately, the change you posted in this comment only gives me empty paths though :(

**Comment by DiegoAE** [comment]
> It's always going to be slow under a failure condition as it has to exhaust every single tile it has access to. The wrong path can be fixed by changing the tail end of `FindPath` to:
> 
> ```
>     CoordinateList path;
>     // We have not got a valid path if our current coordinates do not equal the source
>     if (current->coordinates == source_)
>     {
>         while (current != nullptr) {
>             path.push_back(current->coordinates);
>             current = current->parent;
>         }
>     }
> ```
> 
> This will return an empty path when it fails

@Xenoamor I think you meant `current->coordinates == target_` since that would be the last visited node for a successful path.

**Comment by Xenoamor** [comment]
Yes, well spotted. For some reason when I ported the code I ended up switching my source and target variables around
